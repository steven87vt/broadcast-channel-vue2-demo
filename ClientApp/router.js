import Vue from 'vue'
import VueRouter from 'vue-router'
import { routes } from './routes'

Vue.use(VueRouter)

var base = '/'

let router = new VueRouter({
  mode: 'history',
  base,
  routes: routes
})

export default router
