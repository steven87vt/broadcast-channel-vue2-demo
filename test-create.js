import merge from 'lodash/merge'
import { createStore as createStoreImportInTest } from './ClientApp/store/index'

export function createTestStore(overrides) {
  let http = {
    put: jest.fn(() => {
      Promise.resolve()
    }),
    post: jest.fn(() => {
      Promise.resolve()
    }),
    get: jest.fn(() => {
      Promise.resolve()
    }),
    delete: jest.fn(() => {
      Promise.resolve()
    })
  }

  let store = { plugins: [] }

  if (overrides && typeof overrides.http !== 'undefined' && typeof overrides.http === 'object') {
    merge(http, overrides.http)
  }
  if (overrides && typeof overrides.store !== 'undefined' && typeof overrides.store === 'object') {
    merge(store, overrides.store)
  }

  return createStoreImportInTest(http, store)
}
