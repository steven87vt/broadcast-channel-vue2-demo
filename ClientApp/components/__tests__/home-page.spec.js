import Vuex from 'vuex'
import VueRouter from 'vue-router'
import merge from 'lodash/merge'
import BroadcastChannel from 'broadcast-channel'
import flushPromises from 'flush-promises'
import { shallowMount, createLocalVue } from '@vue/test-utils'

import { createTestStore } from '../../../test-create'
import router from '../../router'
import HomePage from '../home-page'

function createWrapper(overrides) {
  const localVue = createLocalVue()
  localVue.use(Vuex)
  localVue.use(VueRouter)

  const defaults = {
    localVue,
    router,
    store: createTestStore()
  }

  return shallowMount(HomePage, merge(defaults, overrides))
}

function destroyWrapper (wrapper, doneCallback, timeout) {
  wrapper.destroy()
  let time = timeout || 300
  return new Promise((resolve) => {
    setTimeout(() => {
      doneCallback()
      resolve()
    }, time)
  })
}

function onmessageFunc(message) {
  console.log(message)
}

describe('home-page.vue', () => {
  let wrapper
  let wrapper2
  
  let authChannelOnMessageSpy
  let authChannelPostMessageSpy
  let authChannelCloseSpy
  let authChannel2CloseSpy

  let setUserActionSpy
  let userMutationSpy
  
  const userName = 'test user name'
  const expectedUser = {
    name: userName
  }

  beforeAll(async () => {
    jest.setTimeout(9999999)
    await BroadcastChannel.clearNodeFolder()
  })
  afterAll(async (done) => {
    jest.restoreAllMocks()
    await destroyWrapper(wrapper, done, 500)
    await destroyWrapper(wrapper2, done, 500)
  })

  test('home-page matches snapshot test.', async () => {
    wrapper = await createWrapper()
    wrapper2 = await createWrapper()
    expect(wrapper).toMatchSnapshot()

    await wrapper.vm.$nextTick()
    await wrapper2.vm.$nextTick()

    expect(wrapper.vm.authChannel).toBeDefined()
    expect(wrapper2.vm.authChannel).toBeDefined()

    expect(wrapper.vm.user).toBeNull()
    expect(wrapper2.vm.user).toBeNull()
    setUserActionSpy = jest.spyOn(wrapper.vm.$store._actions.setUser, [0])
    userMutationSpy = jest.spyOn(wrapper.vm.$store._mutations.user, [0])
    
    authChannelOnMessageSpy = jest.spyOn(wrapper2.vm.authChannel._addEL.message[0], 'fn') // <-- works
    // authChannelOnMessageSpy = jest.spyOn(wrapper2.vm, 'authChannelMessageHandler') // <-- does not work, probably related to the onmessage invocation never leaving the vm so its not recordable from this level 
    authChannelPostMessageSpy = jest.spyOn(wrapper.vm.authChannel, 'postMessage')
    authChannelCloseSpy = jest.spyOn(wrapper.vm.authChannel, 'close')
    authChannel2CloseSpy = jest.spyOn(wrapper2.vm.authChannel, 'close')
  })

  test('find input, trigger name change', async (done) => {
    
    const nameInput = wrapper.find('input')
    nameInput.element.value = userName
    nameInput.trigger('change')
    
    await new Promise((resolve) => {
      setTimeout(resolve, 1000) // give node time to handle the message event
    }).then(() => {
      expect(wrapper2.vm.user).toBeNull()

      expect(setUserActionSpy).toHaveBeenCalledTimes(1)
      expect(setUserActionSpy).toHaveBeenCalledWith(expectedUser)
  
      expect(userMutationSpy).toHaveBeenCalledTimes(1)
      expect(userMutationSpy).toHaveBeenCalledWith({ user: expectedUser })
  
      expect(authChannelPostMessageSpy).toHaveBeenCalledTimes(1)
      expect(authChannelPostMessageSpy).toHaveBeenCalledWith(expectedUser)
  
      expect(authChannelOnMessageSpy).toHaveBeenCalledTimes(1)
      expect(authChannelOnMessageSpy).toHaveBeenCalledWith(expectedUser)
      done()
    })
  })

  test('destroy wrapper, assert BroadcastChannel#close is executed', async () => {
    wrapper.destroy()
    wrapper2.destroy()
    
    await wrapper.vm.$nextTick()

    expect(authChannelCloseSpy).toHaveBeenCalled()
    expect(authChannel2CloseSpy).toHaveBeenCalled()
  })
})
