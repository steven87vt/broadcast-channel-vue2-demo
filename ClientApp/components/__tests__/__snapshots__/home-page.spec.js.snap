// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`home-page.vue home-page matches snapshot test. 1`] = `
VueWrapper {
  "_emitted": Object {},
  "_emittedByOrder": Array [],
  "isFunctionalComponent": undefined,
}
`;
